sklearn/mixture/gaussian_mixture.py

```diff
@@ -0,0 +1,651 @@
+
+    nk : array-like, shape (n_components,)
+
+    means : array-like, shape (n_components, n_features)
+
+    reg_covar : float
+
+    Returns
+    -------
+    covariances : array, shape (n_components,)
+    """
+    covariances = _estimate_gaussian_covariance_diag(resp, X, nk, means,
+                                                     reg_covar)
+    return covariances.mean(axis=1)
+
+
+def _estimate_gaussian_parameters(X, resp, reg_covar, covariance_type):
+    """Estimate the Gaussian distribution parameters.
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+        The input data array.
+
+    resp : array-like, shape (n_samples, n_features)
+        The responsibilities for each data sample in X.
+
+    reg_covar : float
+        The regularization added to each covariance matrices.
+
+    covariance_type : {'full', 'tied', 'diag', 'spherical'}
+        The type of covariance matrices.
+
+    Returns
+    -------
+    nk : array, shape (n_components,)
+        The numbers of data samples in the current components.
+
+    means : array, shape (n_components, n_features)
+        The centers of the current components.
+
+    covariances : array
+        The sample covariances of the current components.
+        The shape depends of the covariance_type.
+    """
+    compute_covariance = {
+        "full": _estimate_gaussian_covariance_full,
+        "tied": _estimate_gaussian_covariance_tied,
+        "diag": _estimate_gaussian_covariance_diag,
+        "spherical": _estimate_gaussian_covariance_spherical}
+
+    nk = resp.sum(axis=0) + 10 * np.finfo(float).eps
+    means = np.dot(resp.T, X) / nk[:, np.newaxis]
+    covariances = compute_covariance[covariance_type](
+        resp, X, nk, means, reg_covar)
+
+    return nk, means, covariances
+
+
+###############################################################################
+# Gaussian mixture probability estimators
+
+def _estimate_log_gaussian_prob_full(X, means, covariances):
+    """Estimate the log Gaussian probability for 'full' covariance.
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+
+    means : array-like, shape (n_components, n_features)
+
+    covariances : array-like, shape (n_components, n_features, n_features)
+
+    Returns
+    -------
+    log_prob : array, shape (n_samples, n_components)
+    """
+    n_samples, n_features = X.shape
+    n_components = means.shape[0]
+    log_prob = np.empty((n_samples, n_components))
+    for k, (mu, cov) in enumerate(zip(means, covariances)):
+        try:
+            cov_chol = linalg.cholesky(cov, lower=True)
+        except linalg.LinAlgError:
+            raise ValueError("The algorithm has diverged because of too "
+                             "few samples per components. "
+                             "Try to decrease the number of components, or "
+                             "increase reg_covar.")
```

Comment:
I write it here what we discuss earlier to push a trace of your talk on github.
I prefer to not change that function because I should add `reg_covar` as parameter only for the warning message.